include/vkwrap/vkwrap/mman.h

```diff
@@ -0,0 +1,556 @@
+        return format == format_with_stencil;
+    };
+
+    return ranges::any_of( k_formats_with_stencil, isEqualFormats );
+} // hasStencil
+
+inline vk::ImageAspectFlags
+chooseAspectMask( vk::Format format, vk::ImageLayout new_layout )
+{
+    using vk::ImageAspectFlagBits;
+    using vk::ImageAspectFlags;
+    using vk::ImageLayout;
+
+    if ( new_layout != ImageLayout::eDepthStencilAttachmentOptimal )
+    {
+        return ImageAspectFlagBits::eColor;
+    }
+
+    ImageAspectFlags aspect_mask{ ImageAspectFlagBits::eDepth };
+
+    if ( hasStencil( format ) )
+    {
+        aspect_mask |= ImageAspectFlagBits::eStencil;
+    }
+
+    return aspect_mask;
+} // chooseAspectMask
+
+inline std::pair<vk::AccessFlags, vk::AccessFlags>
+chooseAccessMasks( vk::ImageLayout old_layout, vk::ImageLayout new_layout )
+{
+    using vk::AccessFlagBits;
+    using vk::AccessFlags;
+    using vk::ImageLayout;
+
+    AccessFlags src{ AccessFlagBits::eNone };
+    AccessFlags dst{ AccessFlagBits::eNone };
+
+    if ( old_layout == ImageLayout::eUndefined && new_layout == ImageLayout::eTransferDstOptimal )
+    {
+        src = AccessFlagBits::eNone;
+        dst = AccessFlagBits::eTransferWrite;
+    } else if ( old_layout == ImageLayout::eTransferDstOptimal && new_layout == ImageLayout::eShaderReadOnlyOptimal )
+    {
+        src = AccessFlagBits::eTransferWrite;
+        dst = AccessFlagBits::eShaderRead;
+    } else if ( old_layout == ImageLayout::eUndefined && new_layout == ImageLayout::eDepthStencilAttachmentOptimal )
+    {
+        src = AccessFlagBits::eNone;
+        dst = AccessFlagBits::eDepthStencilAttachmentRead | AccessFlagBits::eDepthStencilAttachmentWrite;
+    } else
+    {
+        throw Error{ "chooseAccessMasks: unimplemented layout transition." };
+    }
+
+    return { src, dst };
+} // chooseAccessMasks
+
+inline vk::ImageMemoryBarrier
+createImageBarrierInfo( vk::Image& image, vk::Format format, vk::ImageLayout old_layout, vk::ImageLayout new_layout )
+{
+    using vk::AccessFlagBits;
+    using vk::ImageMemoryBarrier;
+
+    ImageMemoryBarrier barrier{};
+
+    barrier.image = image;
+
+    barrier.oldLayout = old_layout;
+    barrier.newLayout = new_layout;
+
+    /** Any combination of image::sharingMode,
+     * barrier::srcQueueFamilyIndex, barrier::dstQueueFamilyIndex is allowed.
+     */
+    barrier.srcQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+    barrier.dstQueueFamilyIndex = VK_QUEUE_FAMILY_IGNORED;
+
+    auto [ src_access, dst_access ] = chooseAccessMasks( old_layout, new_layout );
+    barrier.srcAccessMask = src_access;
+    barrier.dstAccessMask = dst_access;
+
+    // We use images with one layer and one mipmap.
+    barrier.subresourceRange.baseMipLevel = 0;
+    barrier.subresourceRange.levelCount = 1;
+    barrier.subresourceRange.baseArrayLayer = 0;
+    barrier.subresourceRange.layerCount = 1;
+
+    barrier.subresourceRange.aspectMask = chooseAspectMask( format, new_layout );
```

Comment:
This could be shortened using designated initializers

Reply:
Yep, it is more clear.